File: chaintalker.cpp
#include "chaintalker.hpp"

#include <stdio.h>
#include <string>
#include <iostream>
#include <cpprest/ws_client.h>

using namespace std;
using namespace web;
using namespace web::websockets::client;


websocket_outgoing_message format_send_message(const string& msg) {
    websocket_outgoing_message out_msg;
    out_msg.set_utf8_message(msg);
    cout << msg << endl;
    return out_msg;
}

void chainapi::ChainTalker::open_connection_sync(const string& ws_url) {
    uri address(U(ws_url));
    _client.connect(address).wait();
}

void chainapi::ChainTalker::set_password_sync(const string& pass) {
    // step 1, create password(only need for first call)
    string command = "{\"id\":1,\"method\":\"call\",\"params\":[0,\"set_password\",[\"111111\"]]}";
    _client.send(format_send_message(command)).wait();

    _client.receive().then([](websocket_incoming_message in_msg) {
        return in_msg.extract_string();
    }).then([](string body) {
        cout << body << endl; // or customized logic here to handle error and return
    }).wait();
}

void chainapi::ChainTalker::unlock_sync(const string& pass) {
   // step 2 unlock wallet api
    string command = "{\"id\":1,\"method\":\"call\",\"params\":[0,\"unlock\",[\"111111\"]]}";
    _client.send(format_send_message(command)).wait();

    _client.receive().then([](websocket_incoming_message in_msg) {
        return in_msg.extract_string();
    }).then([](string body) {
        cout << body << endl; // or customized logic here to handle error and return
    }).wait();
}

void chainapi::ChainTalker::import_user_sync(const string& id, const string& key) {
    string command = 
        "{\"id\":1,\"method\":\"call\",\"params\":[0,\"import_key\",[\"" + id + "\", \"" + key + "\"]]}";
    
    _client.send(format_send_message(command)).wait();

    _client.receive().then([](websocket_incoming_message in_msg) {
        return in_msg.extract_string();
    }).then([](string body) {
        cout << body << endl; // or customized logic here to handle error and return
    }).wait();
}

string chainapi::ChainTalker::suggest_brain_key() {
    string command =
        "{\"id\":1,\"method\":\"call\",\"params\":[0,\"suggest_brain_key\",[]]}";

    _client.send(format_send_message(command)).wait();


    _client.receive().then([](websocket_incoming_message in_msg) {
        return in_msg.extract_string();
    }).then([](string body) {
        return body; // or customized logic here to handle error and return
    }).wait();

}

void chainapi::ChainTalker::transfer_sync(const string& from, const string& to, string amount, string memo){
    string command = 
        "{\"id\":1, \"method\":\"call\", \"params\":[0,\"transfer\",[\"" + from + "\", \"" + to + "\", " + amount+", \"BTS\", \""+ memo + "\", true]]}";
    
    _client.send(format_send_message(command)).wait();

    _client.receive().then([](websocket_incoming_message in_msg) {
        return in_msg.extract_string();
    }).then([](string body) {
        cout << body << endl; // or customized logic here to handle error and return
    }).wait();
}

void chainapi::ChainTalker::create_account_with_brain_key_sync(const string& brain_key, const string& account_name) {
    // similar logic as open_connetion.  prepare the command correctly and send ove to ws
    // for api format, plz refer to https://docs.google.com/document/d/1W3KS9Hw_8suqOxqOQBy28-Z_ePBBDI-cRanvQtD_Nfg/edit
    string command =
        "{\"id\":1, \"method\":\"call\", \"params\":[0,\"create_account_with_brain_key\",[\"" + brain_key + "\", \"" + account_name + "\", \"init0\", \"init0\", true]]}";
    _client.send(format_send_message(command)).wait();

    _client.receive().then([](websocket_incoming_message in_msg) {
        return in_msg.extract_string();
    }).then([](string body) {
        cout << body << endl; // or customized logic here to handle error and return
    }).wait();
}

File: python/chaintalker.py
import asyncio
import sys, json
import asyncio
import websockets
import random
import string

class chaintalker:
    async def open(self, url):
        self.websocket = await websockets.connect(url)    
        return self

    async def is_new(self):
        command = '{\"id\":1,\"method\":\"call\",\"params\":[0,\"is_new\",[]]}'
        await self.send(command)
        raw_msg = await self.receive()
        json_msg = json.loads(raw_msg)
        return json_msg['result'] == True

    async def set_password(self, pwd):
        command = '{\"id\":1,\"method\":\"call\",\"params\":[0,\"set_password\",[\"%s\"]]}' % pwd
        await self.send(command)
        raw_msg = await self.receive()
        json_msg = json.loads(raw_msg)
        return json_msg['result'] == None

    async def unlock(self, pwd) :
        command = '{\"id\":1,\"method\":\"call\",\"params\":[0,\"unlock\",[\"%s\"]]}' % pwd
        await self.send(command)
        raw_msg = await self.receive()
        json_msg = json.loads(raw_msg)
        return json_msg['result'] == None
    
    async def import_key(self, account_name_or_id, wif_key) :
        command = '{\"id\":1,\"method\":\"call\",\"params\":[0,\"import_key\",[\"%s\", \"%s\"]]}' % (account_name_or_id, wif_key)
        await self.send(command)
        raw_msg = await self.receive()
        json_msg = json.loads(raw_msg)
        return json_msg['result'] == None
    
    async def suggest_brain_key(self) :
        command = '{\"id\":1,\"method\":\"call\",\"params\":[0,\"suggest_brain_key\",[]]}'
        await self.send(command)
        raw_msg = await self.receive()
        json_msg = json.loads(raw_msg)
        return json_msg['result']['brain_priv_key'], json_msg['result']['wif_priv_key'], json_msg['result']['pub_key']

    async def create_account_with_brain_key(self, brain_key, account_name, registrar_account, referrer_account) :
        command = '{\"id\":1,\"method\":\"call\",\"params\":[0,\"create_account_with_brain_key\",[\"%s\", \"%s\", \"%s\", \"%s\", true]]}' % (brain_key, account_name, registrar_account, referrer_account)
        await self.send(command)
        raw_msg = await self.receive()
        json_msg = json.loads(raw_msg)
        return json_msg['result']

    async def transfer(self, from_account, to_account, amount, memo) :
        command = '{\"id\":1,\"method\":\"call\",\"params\":[0,\"transfer\",[\"%s\", \"%s\", %s, \"BTS\", \"%s\", true]]}' % (from_account, to_account, amount, memo)
        await self.send(command)
        raw_msg = await self.receive()
        json_msg = json.loads(raw_msg)
        return json_msg['result']

    async def send(self, message):
        print("send: " + message)
        await self.websocket.send(message)

    async def receive(self):
        raw_msg = await self.websocket.recv()
        print("receive: " + raw_msg)
        return raw_msg

    def id_generator(self, size=6, chars=string.ascii_lowercase ):
        return ''.join(random.choice(chars) for _ in range(size))
